ragna/ui/central_view.py

```diff
@@ -0,0 +1,681 @@
+            and replace the ChatMessage objects with RagnaChatMessage object.
+        We do it only for the new messages from the rag, not for the existing messages, neither for the messages from the user.
+        """
+
+        def messages_changed(event):
+            if len(chat_interface.objects) != len(self.current_chat["messages"]):
+                return
+
+            assistant = self.current_chat["metadata"]["assistant"]
+            # FIXME: user needs to be dynamic based on the username that was logged in with
+            username = "User"
+
+            needs_refresh = False
+            for i in range(len(chat_interface.objects)):
+                msg = chat_interface.objects[i]
+
+                if not isinstance(msg, RagnaChatMessage) and msg.user != "User":
+                    chat_interface.objects[i] = RagnaChatMessage(
+                        self.current_chat["messages"][i],
+                        username
+                        if self.current_chat["messages"][i]["role"] == "user"
+                        else assistant,
+                        on_click_source_info_callback=self.on_click_source_info_wrapper,
+                    )
+                    msg = chat_interface.objects[i]
+                    needs_refresh = True
+
+            if needs_refresh:
+                chat_interface._chat_log.param.trigger("objects")
+
+        chat_interface.param.watch(
+            messages_changed,
+            ["objects"],
+        )
+
+        # Here, we build a list of RagnaChatMessages from the existing messages of this chat,
+        # and set them as the content of the chat interface
+        chat_interface.objects = self.get_chat_messages()
+
+        # Now that setting all the objects is done, we can watch the change of objects,
+        # ie new messages being appended to the chat. When that happens,
+        # make sure we scroll to the latest msg.
+        chat_interface.param.watch(
+            lambda event: self.param.trigger("trigger_scroll_to_latest"),
+            ["objects"],
+        )
+
+        return chat_interface
+
+    @pn.depends("current_chat", "trigger_scroll_to_latest")
+    def scroll_to_latest_fix(self):
+        """
+        This snippet needs to be re-rendered many times so the scroll-to-latest happens:
+            - each time the current chat changes, hence the pn.depends on current_chat
+            - each time a message is appended to the chat, hence the pn.depends on trigger_scroll_to_latest.
+                trigger_scroll_to_latest is triggered in the chat_interface method, when chat_interface.objects changes.
+
+        Twist : the HTML script node needs to have a different ID each time it is rendered,
+                otherwise the browser doesn't re-render it / doesn't execute the JS part.
+                Hence the random ID.
+        """
+
+        random_id = "".join(
+            random.choice("".join([chr(n) for n in range(65, 91)])) for _ in range(16)
+        )
```

Comment:
That looks rather complicated to just get a random ID. What about

```suggestion
        random_id = str(uuid.uuid4())
```